text/0069-api-v2.md

```diff
@@ -0,0 +1,224 @@
+3. Disable `API V2` in TiKV config file and restart TiKV (User should take the responsibility to offline all tikv clients excluding TiDB). Or set by online config change API (Not proposed in this RFC, but is good to have).
+
+#### Data migration
+
+It's reasonable to provide a way to import and export non-TiDB data in TiKV during the upgrade or downgrade. On TiKV before 4.0, the only way to do that is `scan` and `batch_put` on the client. After 4.0, TiKV start to support importing SST file into TxnKV, and after 5.1, importing on RawKV is also supported. You can find more information in [`RFC: Online Bulk Load for RawKV`](https://github.com/tikv/rfcs/pull/72).
+
+### Implementation Details
+
+#### PD
+
+Add the new APIs described [above](#Keyspace-Management).
+
+#### TiKV Server
+
+In TiKV config file, add a new configuration `storage.api_version`. When enabled, `storage.enable_ttl` must also be enabled.
+
+In kvdb, add a store meta `api_version`. When the store meta mismatches the config `storage.enable_ttl`, it means that the user is switching the API version, then check no non-TiDB exist, and then save the new api version in store meta.
+
+In kvproto message `SSTMeta`, add `api_version`. Reject the SST file if the version is mismatched.
+
+In TiKV gRPC's context, add a field `api_version`.
+
+If `storage.api_version=2`:
+
+- Run TTL compaction filter only on the keys that start with RawKV prefix.
+
+- Use the `API V2` Value encode in `RawStore`, `TTLStore` and `sst_importer`.
+
+- If the request's context has `api_version=1`:
+  - Reject the request unless it's a TxnKV request and the keys starting with `m` or `t`.
+
+- If the request's context has `api_version=2`:
+  - Only allow the key that has RawKV prefix for RawKV requests.
+  - Only allow the key that has TxnKV prefix for TxnKV requests.
+
+If `storage.api_version=1`:
+
+- Reject all requests with `api_version=2` in the context.
+
+#### TiKV Client
+
+Provide two modes for users:
+
+- V2:
+  - Fetch keyspace prefix by keyspace name from PD and then prepend `k{keyspace prefix}x` on TxnKV keys or prepend `k{keyspace prefix}r` on RawKV keys.
+  - Set `api_version=2` in TiKV gRPC's `Context`.
+  - Disallow specify CF in `RawClient`.
+  - Allow user to specify a keyspace for a session of `RawClient` or `TxnCient`. Default keyspace is named `default`.
+  - Fetch keyspace information from PD every 5 mins. Destory the client session if the keyspace it's using is deleted.
+
+- V1:
+  - Behaves jusk like current client.
+  - Set `api_version=1` in TiKV gRPC's `Context`.
+
+Listed below is the compatibility matrix:
+
+|              | V1 Server | V2 Server |
+| ------------ | --------- | --------- |
+| V1 RawClient | Raw Data  | Forbidden |
+| V1 TxnClient | Txn Data  | TiDB Data |
+| V2 RawClient | Forbidden | Raw Data  |
+| V2 TxnClient | Forbidden | Txn Data  |
+
+### CDC / BR
+
+Since all access to TiDB is unchanged during the upgrade, CDC and BR should work the same after upgrade/downgrade.
```

Comment:
/cc @overvenus PTAL

Reply:
What do you mean by the `key codec in the TiKV read/write`?

Reply:
For example, parsing keys or constructing keys in coprocessors.

Reply:
coprocessor only read and write tidb data, and since tidb data will not change, coprocessor will not change either.

Reply:
So tidb will not use API V2?

Reply:
tidb uses api v1